File: December2021/RangeSumofBST.java
// TC : O(n)
// SC : O(logn)
class Solution {
	private int sum = 0;
	public int rangeSumBST(TreeNode root, int L, int R) {
		rangeSumBSTHelper(root, L, R);
		return sum;
	}
	private void rangeSumBSTHelper(TreeNode root, int L, int R){
		if(root==null){
			return ;
		}

		if(root.val>=L && root.val<=R){
			sum+=root.val;
			rangeSumBSTHelper(root.left, L, root.val);
			rangeSumBSTHelper(root.right, root.val, R);

		} else if(root.val<L){
			rangeSumBSTHelper(root.right, L, R);
		} else if(root.val>R){
			rangeSumBSTHelper(root.left, L, R);
		}

	}
}

File: November2022/C++/range-sum-of-bst.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
private:
    void findRangeSum(int &totalSum, int low, int high, TreeNode*root){
        
        if(root==NULL){
            return;
        }
        
        if(root->val<low){
            
          findRangeSum(totalSum,low,high,root->right);
        }
        
          else if(root->val>high){
            
          findRangeSum(totalSum,low,high,root->left);
        }
        
        else{
        totalSum += root->val;
        
       findRangeSum(totalSum,low,high,root->left);
           findRangeSum(totalSum,low,high,root->right);
            
        }
    }
public:
    
    int rangeSumBST(TreeNode* root, int low, int high) {
        
        int totalSum = 0;
        
        findRangeSum(totalSum,low, high, root);
        
        return totalSum;
    }
};